src/undo_log/undo_log.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2018-2022 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include "../value.h"
+#include "../graph/entities/node.h"
+#include "../graph/entities/edge.h"
+#include "../execution_plan/ops/shared/update_functions.h"
+
+// this file contains the rollback functionality:
+// when a query finished in timeout or exception
+// all it's commited data should be rollbacked.
+// the UndoLog accumulates the commited changes of the query.
+
+
+// the type of operation of each item in the undo_list
+typedef enum {
+	UL_UPDATE_NODE = 0,   // undo node update
+	UL_UPDATE_EDGE,       // undo edge update
+	UL_CREATE_NODE,       // undo node creation
+	UL_CREATE_EDGE,       // undo edge creation
+	UL_DELETE_NODE,       // undo node deletion
+	UL_DELETE_EDGE        // undo edge deletion
+} UndoOpType;
+
+// the abstract type of items in the undo_list
+typedef struct UndoOp {
+	union {
+		Node n;
+		Edge e;
+	} entity;
+	union {
+		struct {
+			LabelID *labels;
+			uint label_count;
+		} delete_node;
+		struct {
+			Attribute_ID attr_id;
+			SIValue orig_value;
+		} update;
+	} data;
+	UndoOpType type;
+} UndoOp;
```

Comment:
A bit unclear how all type of UNDOs fit into this struct, either create a struct per operation, or rename and add comments